plugin/plugin.go

```diff
@@ -1,243 +1,152 @@
+// Package plugin provides the toolings to use the notation plugin.
+//
+// includes a CLIManager and a CLIPlugin implementation.
 package plugin
 
 import (
+	"bytes"
 	"context"
-	"time"
-)
-
-// Prefix is the prefix required on all plugin binary names.
-const Prefix = "notation-"
-
-// ContractVersion is the <major>.<minor> version of the plugin contract.
-const ContractVersion = "1.0"
-
-// Command is a CLI command available in the plugin contract.
-type Command string
-
-const (
-	// CommandGetMetadata is the name of the plugin command
-	// which must be supported by every plugin and returns the
-	// plugin metadata.
-	CommandGetMetadata Command = "get-plugin-metadata"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"path/filepath"
 
-	// CommandDescribeKey is the name of the plugin command
-	// which must be supported by every plugin that has the
-	// SIGNATURE_GENERATOR.RAW capability.
-	CommandDescribeKey Command = "describe-key"
-
-	// CommandGenerateSignature is the name of the plugin command
-	// which must be supported by every plugin that has the
-	// SIGNATURE_GENERATOR.RAW capability.
-	CommandGenerateSignature Command = "generate-signature"
-
-	// CommandGenerateEnvelope is the name of the plugin command
-	// which must be supported by every plugin that has the
-	// SIGNATURE_GENERATOR.ENVELOPE capability.
-	CommandGenerateEnvelope Command = "generate-envelope"
-
-	// CommandVerifySignature is the name of the plugin command
-	// which must be supported by every plugin that has
-	// any SIGNATURE_VERIFIER.* capability
-	CommandVerifySignature Command = "verify-signature"
+	"github.com/notaryproject/notation-go/plugin/proto"
 )
 
-// Capability is a feature available in the plugin contract.
-type Capability string
+var executor commander = &execCommander{} // for unit test
 
-// In returns true if the Capability is present in the given array of capabilities
-func (c Capability) In(capabilities []Capability) bool {
-	for _, capability := range capabilities {
-		if c == capability {
-			return true
-		}
-	}
-	return false
+// GenericPlugin is the base requirement to be an plugin.
+type GenericPlugin interface {
```

Comment:
`GenericPlugin` gives an impression that you can get a plugin just by implementing `GetMetadata` method, but that is not correct. A plugin needs to implement SignPlugin methods or VerifyPlugin methods to truly become a functional plugin. That being said, I suggest we rename this to `pluginBase` (with a small `p`). I don't think we need to export this interface to the library consumers.

Reply:
According to our plugin spec, you don't need to implement signing or verification capability to be a plugin. As long as the GetMetadata capability is implemented, it can be identified as a plugin. `pluginBase` is more like an implementation of the interface. So, I think `genericPlugin` is better.

Reply:
Oh, Shiwei remind me that a external interface includes a internal interface is not a convention in Golang, so maybe `GenericPlugin` is better.

Reply:
> As long as the GetMetadata capability is implemented, it can be identified as a plugin.

How useful that plugin is going to be? Without sing or verify methods that plugin is useless. I suggest we make `GenericPlugin` private or remove entirely, until we see a need for exporting it.

Reply:
I also considered to make it private, but if it is private, that is to say, we can change the method of a private interface without notice, which is not a breaking change. In face, we should keep the `GenericPlugin` being public, and any change to it should be a breaking change.

Also, removing it will cause duplicated GetMetadata() definition.

Can we keep the original name?